### DISCUSSION TITLE
Error in MongoDBUserDatabase: Future exception was never retrieved

### DISCUSSION THREAD

**Discussion by krishnadubba** [fastapi-users/fastapi-users#634] [Q&A] [answered]
Hi all,
I am getting this error with fastapi-users. I narrowed it down to MongoDBUserDatabase. Did anyone encounter this before? Any help would be appreciated. My "motor" version is 2.3.1 and I am using CosmosDB with Mongodb api.

Future exception was never retrieved
future: <Future finished exception=OperationFailure("collation, full error: {'ok': 0.0, 'errmsg': 'collation', 'code': 115, 'codeName': 'CommandNotSupported'}")>
Traceback (most recent call last):
  File "\lib\concurrent\futures\thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "\lib\site-packages\pymongo\collection.py", line 2059, in create_index
    return self.__create_indexes([index], session, **cmd_options)[0]
  File "\lib\site-packages\pymongo\collection.py", line 1949, in __create_indexes
    self._command(
  File "\lib\site-packages\pymongo\collection.py", line 238, in _command
    return sock_info.command(
  File "\lib\site-packages\pymongo\pool.py", line 683, in command
    return command(self, dbname, spec, slave_ok,
  File "\lib\site-packages\pymongo\network.py", line 159, in command
    helpers._check_command_response(
  File "\lib\site-packages\pymongo\helpers.py", line 164, in _check_command_response
    raise OperationFailure(errmsg, code, response, max_wire_version)
pymongo.errors.OperationFailure: collation, full error: {'ok': 0.0, 'errmsg': 'collation', 'code': 115, 'codeName': 'CommandNotSupported'}

Thanks.
Krishna

**Comment by frankie567** [comment, marked as answer]
It looks like it's related to the creation of the e-mail index with case-insensitive collation:

https://github.com/frankie567/fastapi-users/blob/5b76d5d90a7e47bb6faefa337715f13a9d37f3bb/fastapi_users/db/mongodb.py#L33-L42

From what I see, CosmosDB emulates parts of the MongoDB API and it seems that they don't support case-insensitive indices: https://docs.microsoft.com/en-us/azure/cosmos-db/mongodb-feature-support-40#index-properties

It's quite a corner-case here. The only solution I see for you is to sub-class `MongoDBUserDatabase` and overload `__init__` and `get_by_email` to remove all mentions of the email collation. The only drawback is that your application will be case-sensitive when checking for e-mail (`king.arthur@camelot.bt` and `King.Arthur@camelot.bt` will be two different e-mails).
